### PR TITLE
feat: added progress bar based on repo size/downloaded bytes

### PR DESCRIPTION
**Pull request by mooyg**
Added progress bar based on github repo size and used a package [git2-rs](https://github.com/rust-lang/git2-rs) for the helper functions

### CODE REVIEW

**Review by mintbomb27 — CHANGES_REQUESTED** [review]
Hey @mooyg . Please try building your code as well when you try to resolve merge conflicts. The code is kinda messed up, and the new code from `master` had a slightly different implementation as it fetches `submodules.json` from the repo itself.

Please see the new implementation and try to accommodate your changes with it :)

File: submodules.json [hunk 1]
```diff
@@ -17,9 +26,14 @@
...
+    "tags": ["jquery", "jqueryjs", "js"],
+    "owner": "jquery-boilerplate"
   },
   {
+    "name": "android-boilerplate",
+    "url": "https://github.com/ribot/android-boilerplate",
+    "tags": ["android", "mvvm", "rxjava"],
+    "owner": "ribot"
     "name": "nginx-boilerplate",
     "url": "https://github.com/nginx-boilerplate/nginx-boilerplate",
```
Comment (Mannan-Goyal) [inline submodules.json:38]: There are duplicate values and I don't think you need to add a separate field for owner as you can just parse it from the URL. For example, if the URL is this - https://github.com/IEEE-VIT/templa-rs the repo owner can be parsed from this string itself.
You can just take 1 argument and take the last part of the URL i.e., `IEEE-VIT/templa-rs` and query the API with that

File: src/main.rs [hunk 2]
```diff
@@ -4,6 +4,10 @@ use serde_json::Value;
 
 mod command;
 mod models;
+mod parse_json;
+mod search;
+mod size;
+mod tui;
 mod search;
```
Comment (Mannan-Goyal) [inline src/main.rs:11]: Modules are listed twice

File: src/tui.rs [hunk 3]
```diff
@@ -21,6 +22,7 @@ use tui::{
     Terminal,
 };
 
+pub fn render_tui(key: String, proj_name: &str) {
```
Comment (Mannan-Goyal) [inline src/tui.rs:25]: Submodules are being passed as argument in main by being dynamically fetched from the remote repository but, you haven't taken that as argument

File: src/tui.rs [hunk 4]
```diff
@@ -79,6 +82,9 @@ pub fn render_tui(key: String, proj_name: &str, submodules: &[Submodule]) {
                 }
                 KeyCode::Down => {
                     if let Some(selected) = template_list_state.selected() {
+                        let templates_length = parse_json::read_json()
+                            .expect("can fetch template list")
+                            .len();
                         let templates_length = submodules.len();
```
Comment (Mannan-Goyal) [inline src/tui.rs:88]: Templates Length repeated again, just take care of repetition generally in the whole codebase

File: src/models/structs.rs [hunk 5]
```diff
@@ -5,4 +5,5 @@ pub struct Submodule {
     pub name: String,
     pub url: String,
     pub tags: Vec<String>,
+    pub owner: String,
```
Comment (Mannan-Goyal) [inline src/models/structs.rs:8]: Not needed

**Review by Mannan-Goyal — CHANGES_REQUESTED** [review]
Please see the comments

**Pull request closed without merging**